prusti-interface/src/specs/typed.rs

```diff
@@ -96,96 +95,74 @@ pub enum SpecificationItem<T> {
 }
 
 impl<T> SpecificationItem<T> {
+    pub fn is_empty(&self) -> bool {
+        matches!(self, SpecificationItem::Empty)
+    }
+
+    /// Extract the value of this item via different refinement strategies
+    pub fn extract_refinements(&self) -> ExtractedRefinement<T> {
+        ExtractedRefinement { item: self }
+    }
+
     /// Returns the contained value of this item
-    pub fn get(&self) -> Option<(Option<&T>, &T)> {
+    fn get(&self) -> Option<(Option<&T>, &T)> {
         match self {
             SpecificationItem::Empty => None,
             SpecificationItem::Inherited(val) => Some((None, val)),
             SpecificationItem::Inherent(val) => Some((None, val)),
-            SpecificationItem::Refined(from, to) => Some((Some(from), to))
-        }
-    }
-
-    pub fn get_or<'a>(&'a self, default: &'a T) -> (Option<&'a T>, &'a T) {
-        self.get().unwrap_or((None, default))
-    }
-
-    pub fn is_empty(&self) -> bool {
-        matches!(self, SpecificationItem::Empty)
-    }
-}
-
-impl<T> SpecificationItem<Vec<T>> {
-    pub fn iter(&self) -> SpecItemIter<'_, T> {
-        if self.is_empty() {
-            SpecItemIter::empty()
-        } else {
-            SpecItemIter::new(self.get().map(|(_, val)| val).unwrap())
+            SpecificationItem::Refined(from, to) => Some((Some(from), to)),
         }
     }
 }
 
-pub struct SpecItemIter<'a, T> {
-    items: Option<&'a Vec<T>>,
-    pos: usize,
+/// A type to access possibly refined [SpecificationItem] values with different strategies
+pub struct ExtractedRefinement<'a, T> {
```

Comment:
Why this extra wrapper? Can the "refinement strategies" not be defined directly on `SpecificationItem`?

Reply:
They could be directly on `SpecificationItem`. The wrapper makes it more readable imo.

Reply:
Not sure I agree :) I would especially like to avoid the call to `.extract_refinements()` at the call sites.

Reply:
I personally like the idea to make explicit that we are extracting information from a `SpecificationItem`, i.e. preparing it to be used for encoding. 

But this is of course just cosmetics, these methods can all life on `SpecificationItem` directly, I guess this is what you suggest.

Reply:
I've moved the methods to `SpecificationItem`